gui: Bilingual GUI error messages

hebasto opened the pull request:
This is an alternative to #15340 (it works with the `Chain` interface; see: https://github.com/bitcoin/bitcoin/pull/15340#issuecomment-502674004).
Refs:
- #16218 (partial fix)
- https://github.com/bitcoin/bitcoin/pull/15894#issuecomment-487947077

This PR:
- makes GUI error messages bilingual: user's native language + untranslated (i.e. English)
- insures that only untranslated messages are written to the debug log file and to `stderr` (that is not the case on master).

If a translated string is unavailable only an English string appears to a user.

Here are some **examples** (updated):

![Screenshot from 2020-04-24 17-08-37](https://user-images.githubusercontent.com/32963518/80222043-e2458780-864e-11ea-83fc-197b7121dba5.png)

![Screenshot from 2020-04-24 17-12-17](https://user-images.githubusercontent.com/32963518/80222051-e5407800-864e-11ea-92f7-dfef1144becd.png)

* `qt5ct: using qt5ct plugin` message is my local environment specific; please ignore it.

---

Note for reviewers: `InitWarning()` is out of this PR scope.

hebasto commented:
@MarcoFalke your https://github.com/bitcoin/bitcoin/pull/15340#discussion_r294111089 has been addressed.

ryanofsky commented:
Concept ACK, especially for the code cleanup making error handling more consistent. Some thoughts:

- I'm not sure it would be better to show same message twice in error dialogs, than to only show the translated message and something like "More detailed debug information can be found in debug.log," or maybe just the translated message and a "View debug log..." button, to reduce the amount of clutter. I'd hope that in most cases translated error messages would be clear enough to help users solve problems, and only more rarely would users need untranslated strings for to report a bug or look things up.

- Using `_("")` syntax for translated strings is pretty common and well understood, but `_("", true)` for bilingual strings seems harder to distinguish and figure out. Might worth considering other syntaxes `_b("")`, `bilingual("")`, `translate("")` or even `""_translate` with [user literal](https://en.cppreference.com/w/cpp/language/user_literal) syntax. Other option would be not having any special syntax and just adding normal functions that return literal and formatted `bilingual_str` objects.

PR looks good in its current form though, and it would be good to get opinions from other reviewers.

MarcoFalke commented:
Is there even a point in having two different translation functions long term? Shouldn't all non-gui code use the one that returns a `bilingual_str`?

MarcoFalke commented:
Is there even a point in having two different translation functions long term? Shouldn't all non-gui code use the one that returns a `bilingual_str`?

ryanofsky commented:
> Is there even a point in having two different translation functions long term? 

Good point. I don't think there is. So another alternative could be to update the `_` function to return both translated and untranslated strings, and not have any new syntax at all.

hebasto commented:
Rebased.

practicalswift commented on src/util/translation.h:
```diff
@@ -0,0 +1,50 @@
+template <typename... Args>
+bilingual_str format(const bilingual_str& fmt, const Args&... args)
+{
+    return std::move(bilingual_str{format(fmt.original, args...), format(fmt.translated, args...)});
```
I think `std::move` prevents copy elision here.

hebasto replied:
Fixed.

practicalswift commented on src/util/translation.h:
```diff
@@ -0,0 +1,50 @@
+ */
+inline bilingual_str _(const char* psz, bool ignored)
+{
+    return std::move(bilingual_str{psz, G_TRANSLATION_FUN ? (G_TRANSLATION_FUN)(psz) : psz});
```
Same here :-)

hebasto replied:
Fixed.

hebasto commented:
Rebased.

@practicalswift Thank you for review.
Your comments have been addressed. Nothing prevents return value optimization now ;)

hebasto commented:
@ryanofsky 
> Using `_("")` syntax for translated strings is pretty common and well understood, but `_("", true)` for bilingual strings seems harder to distinguish and figure out. Might worth considering other syntaxes `_b("")`, `bilingual("")`, `translate("")` or even `""_translate` with user literal syntax.

IIUC, this will require appropriate changes for the `share/qt/extract_strings_qt.py` script, right?

MarcoFalke commented on src/util/translation.h:
```diff
@@ -0,0 +1,50 @@
+ * Translation function.
+ * If no translation function is set, simply return the input.
+ */
+inline std::string _(const char* psz)
```
What about this?

```suggestion
inline bilingual_str _(const char* psz)
```

Have you seen my comment here: https://github.com/bitcoin/bitcoin/pull/16224#issuecomment-502737888?

hebasto replied:
> Have you seen my comment here: #16224 (comment)?

Sure!

IIUC, your suggestion implies some kind of type conversion (from `bilingual_str` to `std::string`) for arguments of some functions and in assignment statements?

MarcoFalke replied:
If you don't need the original string, it can be dropped with `_().translated`, no? I understand that the changes are more disruptive, but it seems clearer than the unused optional argument.

It could probably be done as a scripted-diff.

hebasto commented:
@MarcoFalke 
> Is there even a point in having two different translation functions long term? Shouldn't all non-gui code use the one that returns a `bilingual_str`?

Implemented.

> It could probably be done as a scripted-diff.

Done.

Also rebased on top of #16278.

MarcoFalke commented:
ACK 72d3599e5b06fb1799c8b3689e76d7b97b067531

Not sure about the huge scripted diff, you can probably replace the body with 

```
sed -i --regexp-extended 's/\b_\("([^"]+|\\")*"\)/&.translated/g' $(git grep --files-with-matches '\b_(' src)
```

hebasto commented:
@MarcoFalke Thank you for your review.
> Not sure about the huge scripted diff, you can probably replace the body...

Done.

hebasto commented:
@MarcoFalke Thank you for your review.

> Not sure about the huge scripted diff, you can probably replace the body with
> ```
> sed -i --regexp-extended 's/\b_\("([^"]+|\\")*"\)/&.translated/g' $(git grep --files-with-matches '\b_(' src)
> ```

Done:
```
sed -i 's/\b_("\([^"]\|\\"\)*")/&.translated/g' $(git grep --files-with-matches '\b_("' src)
```

hebasto commented:
GH rearranged this PR commits. The amended commit 54419c82dd025df1b214ac3b99f85d66f3b9b403 became the last in the https://github.com/bitcoin/bitcoin/pull/16224/commits.

Is this an issue? If so, how may I fix it?

promag commented:
> Is this an issue? If so, ~how may I fix it~ should I fix the commit order?

@hebasto no, it's fine.

promag commented:
Concept ACK, please don't merge this yet.

promag commented on src/init.cpp:
```diff
@@ -745,17 +745,17 @@ static void ThreadImport(std::vector<fs::path> vImportFiles)
-        return false;
+    if (!ECC_InitSanityCheck()) {
+        std::string rare_error_do_not_translate = "Elliptic curve cryptography sanity check failure. Aborting.";
+        return InitError(bilingual_str{rare_error_do_not_translate, rare_error_do_not_translate});
```
e985a077f5f6b5e154f2cb4be25a50a45e00e211

These changes are unnecessary if you keep `error InitError(const std::string&)` and just do:
```cpp
bool InitError(const std::string& str)
{
    return InitError(bilingual_str{str, str});
}
```

hebasto replied:
The function with signature `InitError(const std::string&)` has been removed intentionally to prevent emerging of untranslated error messages in the future. All such cases should be implemented in the code explicitly.

promag commented on src/ui_interface.cpp:
```diff
@@ -54,13 +54,22 @@ void CClientUIInterface::NotifyBlockTip(bool b, const CBlockIndex* i) { return g
 
+bool InitError(const bilingual_str& str)
+{
+    bilingual_str bs{str.original, str.translated};
```
a9641a7ac12e3eae763faadbf9bfeeb5f14e3717

Just
```cpp
bilingual_str bs{str};
```

hebasto replied:
Agree. Will be fixed tonight.

hebasto replied:
Done.

hebasto replied:
It fails on trusty (https://travis-ci.org/bitcoin/bitcoin/jobs/554836240).

promag replied:
iono about that error 😕 somehow it doesn't use copy constructor.

hebasto replied:
[Copy initialization](https://en.cppreference.com/w/cpp/language/copy_initialization) will do:
```suggestion
    bilingual_str bs = str;
```

promag reviewed (commented):
Agree with @ryanofsky, I think original text could be set with `QMessageBox::setDetailedText` and `BitcoinGUI::message` could be extended to accept that parameter.

hebasto commented:
@promag Thank you for your review.
> Agree with @ryanofsky, I think original text could be set with `QMessageBox::setDetailedText` and `BitcoinGUI::message` could be extended to accept that parameter.

Do you think the `Show Details...` button on an error message window is expressive enough for a user? Maybe `QMessageBox::informativeText`?

promag commented:
I don't think so, the idea is to allow access to the original message while not displaying both.

hebasto commented:
@promag 
> I don't think so, the idea is to allow access to the original message while not displaying both.

In this case "Original message:" prefix seems redundant in the detailed text, right?

promag commented:
> In this case "Original message:" prefix seems redundant in the detailed text, right?

I don't mind the prefix, actually it can help later if we decide to add more details, like:
```
Original message: This is a bad message in english.
Erro code: 123
Location: some_file.cpp:line
```

hebasto commented:
@promag 
> I don't think so, the idea is to allow access to the original message while not displaying both.

Implemented.

promag commented:
On macos
![Screenshot 2019-07-07 at 00 17 31](https://user-images.githubusercontent.com/3534524/60761979-abb16880-a04c-11e9-8286-110d0d64a45b.png)
![Screenshot 2019-07-07 at 00 17 34](https://user-images.githubusercontent.com/3534524/60761980-abb16880-a04c-11e9-89d5-4c7f3f8feb92.png)

hebasto commented:
Rebased.

MarcoFalke commented:
Could split all the refactoring-only up to (and including) 4e1eda717ed282f7a8ca67329d5a0669b6832ef8 to a separate pull request?

hebasto commented:
@MarcoFalke 
> Could split all the refactoring-only up to (and including) [4e1eda7](https://github.com/bitcoin/bitcoin/commit/4e1eda717ed282f7a8ca67329d5a0669b6832ef8) to a separate pull request?

Done: #16362

hebasto commented:
Rebased.

ryanofsky commented on src/init.cpp:
```diff
@@ -1054,7 +1058,7 @@ bool AppInitParameterInteraction()
         const std::string minChainWorkStr = gArgs.GetArg("-minimumchainwork", "");
         if (!IsHexNumber(minChainWorkStr)) {
-            return InitError(strprintf("Invalid non-hex (%s) minimum chain work value specified", minChainWorkStr));
+            return InitError(strprintf(_("Invalid non-hex (%s) minimum chain work value specified"), minChainWorkStr));
```
This change as a whole seems to be adding a lot of new translated strings. But I wonder if it would be better to be more conservative and not add new translations here, instead saving that part for a followup PR and selectively leaving less common errors untranslated. One way to do this might be with a `DoNotTranslate` function

```c++
bilingual_str DoNotTranslate(std::string original)
{
    return {original, original};
}
```

hebasto replied:
`DoNotTranslate()` has been implemented.

Sjors replied:
Nit: don't forget to actually use `Untranslated ` here.

ryanofsky reviewed (approved):
utACK 32cf78af45c662617eaf91bb3bb07e27151a3a95. I skimmed all the commits, but found it most easy just to review the final consolidated diff, because some of intermediate commits update the same lines more than once.

This change seems acceptable in its current state, but my preference would be to have a new refactoring-only PR that introduces bilingual string instances _without_ changing behavior. The refactoring-only PR could:

1. Introduce the bilingual_str class
1. Change `_` to return bilingual strings, and change `MessageBox` and `InitError` (and maybe `AbortNode` and `InitWarning`) to accept bilingual strings
1. Add `DoNotTranslate` (see below) calls wherever needed to maintain existing behavior and avoid breaking compilation.

After that PR, this PR altering the GUI and adding new translations could be smaller and easier to understand.

hebasto commented:
Rebased.

hebasto commented:
@ryanofsky Thank you for your review.

> The refactoring-only PR could:
> ...
> 2. Change `_` to return bilingual strings, and change `MessageBox` and `InitError` (and maybe `AbortNode` and `InitWarning`) to accept bilingual strings

Such commit could be very complicated. Currently, this PR introduces bilingual abilities to the related functions (`ThreadSafeMessageBox`, `InitError`) commit by commit.

Also this PR has been already [split](https://github.com/bitcoin/bitcoin/pull/16224#issuecomment-509745808) as @MarcoFalke suggested.

hebasto commented on src/init.cpp:
```diff
@@ -956,7 +960,7 @@ bool AppInitParameterInteraction()
     // Warn if unrecognized section name are present in the config file.
     for (const auto& section : gArgs.GetUnrecognizedSections()) {
-        InitWarning(strprintf("%s:%i " + _("Section [%s] is not recognized."), section.m_file, section.m_line, section.m_name));
+        InitWarning(strprintf("%s:%i " + _("Section [%s] is not recognized.").translated, section.m_file, section.m_line, section.m_name));
```
You are reading my mind ;)
`InitWarning` has been left for a separate PR.

hebasto commented on src/init.cpp:
```diff
@@ -1622,9 +1626,10 @@ bool AppInitMain(InitInterfaces& interfaces)
-                    strLoadError + ".\n\n" + _("Do you want to rebuild the block database now?"),
-                    strLoadError + ".\nPlease restart with -reindex or -reindex-chainstate to recover.",
+                    bilingual_str{strLoadError.original + "\n\n" + bs.original, strLoadError.translated + "\n\n" + bs.translated},
+                    strLoadError.original + ".\nPlease restart with -reindex or -reindex-chainstate to recover.",
```
> ... is it intentional that only the first part of the translated string is actually translated, and the second half is not?

Yes, it is. The second half is a non-interactive message which is printed to the `debug.log` and `std::cerr`:
https://github.com/bitcoin/bitcoin/blob/18c3521c2c72cee05094e49ec97c4ef8facdd1d8/src/noui.cpp#L47-L50

hebasto commented:
@ryanofsky Added TODO comments and the `DoNotTranslate()` function.

hebasto commented on src/validation.cpp:
```diff
@@ -1382,7 +1383,7 @@ static bool AbortNode(const std::string& strMessage, const std::string& userMess
     LogPrintf("*** %s\n", strMessage);
     if (!userMessage.empty()) {
-        uiInterface.ThreadSafeMessageBox(userMessage, "", CClientUIInterface::MSG_ERROR | prefix);
+        uiInterface.ThreadSafeMessageBox(bilingual_str{userMessage, userMessage}, "", CClientUIInterface::MSG_ERROR | prefix);
```
Some TODO comments have been added.

hebasto commented:
Rebased on top of #16355.

hebasto commented:
@promag 
> ... please don't merge this yet.

Could we move on? Would you mind looking into #16362 too?

hebasto commented:
Rebased.

hebasto commented:
@MarcoFalke 
Rebasing on top of #16366 caused a bilingual representation of some messages: [diff](https://github.com/bitcoin/bitcoin/compare/be65bb641de05720a4eb9bd854b0f53bb58c255b..af608a0cf08e742fef1e5b0006de91c1db18f781).

I suppose some of them should be untranslated intentionally. If so, let me know to make changes, please?

ryanofsky commented:
> I suppose some of them should be untranslated intentionally. If so, let me know to make changes, please?

There could be other opinions, but I'd recommend replacing `_` with `DoNotTranslate` here to avoid refactoring and changing behavior in the same PR. (Sorry for sounding like a broken record on mixing refactoring and behavior changes, but I think if we want to add new translations, this should happen in a small separate standalone PR adding new translations, not mixed in with everything else here).

MarcoFalke commented on src/bitcoind.cpp:
```diff
@@ -19,6 +19,9 @@
 #include <util/strencodings.h>
 #include <util/system.h>
 #include <util/threadnames.h>
+#include <util/translation.h>
```
translations are only for the gui, this should never be added here

ryanofsky replied:
> translations are only for the gui, this should never be added here

I haven't looked at the context for this comment yet, but in general translations are desirable when writing error messages to stderr, right? My understanding is we only want to avoid translating very technical messages, rare messages, and messages sent to debug.log.

hebasto replied:
> translations are only for the gui, this should never be added here

That is correct. But without `#include <util/translation.h>` the linker fires an error:
```
  CXXLD    bitcoind
libbitcoin_server.a(libbitcoin_server_a-init.o): In function `_(char const*)':
/home/hebasto/github/bitcoin/src/./util/translation.h:39: undefined reference to `G_TRANSLATION_FUN[abi:cxx11]'
```

MarcoFalke commented on src/qt/bitcoin.cpp:
```diff
@@ -459,7 +459,7 @@ int GuiMain(int argc, char* argv[])
+        node->initError(strprintf(_("Error parsing command line arguments: %s\n"), error));
         // Create a message box, because the gui has neither been created nor has subscribed to core signals
         QMessageBox::critical(nullptr, PACKAGE_NAME,
             // message can not be translated because translations have not been initialized
```
:eyes: Please remove the `_(` three lines above.

hebasto replied:
Correct. Fixed.

MarcoFalke commented on src/qt/bitcoin.cpp:
```diff
@@ -521,7 +521,7 @@ int GuiMain(int argc, char* argv[])
         node->selectParams(gArgs.GetChainName());
     } catch(std::exception &e) {
-        node->initError(strprintf("%s\n", e.what()));
+        node->initError(strprintf(_("%s\n"), e.what()));
```
Here and the two other instances: The The gui is not yet subscribed to core signals and can't pick up the translation, which is why it is manually translated below.

hebasto replied:
Correct. Fixed.

hebasto commented:
The latest push:
- rebased on top of #16362 
- fixed rebasing on top of #16366: no more translations in `bitcoind` and `Node::initError()`
- forward declaration of `struct bilingual_str`

hebasto commented:
Rebased on top of 367b023ae444e7d9641c4a3469f9154461e50e68.

hebasto commented:
Rebased after #16620 has been merged.

hebasto commented:
Rebased.

hebasto commented:
Rebased.

laanwj commented:
Code review ACK 322ffab8771a2a968246deb32750e1b6b1b930b4

hebasto commented:
Rebased after #16743 has been merged.

hebasto commented:
@MarcoFalke @ryanofsky @practicalswift @promag 

Would you mind re-reviewing this PR?

MarcoFalke commented on src/init.cpp:
```diff
@@ -1619,7 +1619,8 @@ bool AppInitMain(InitInterfaces& interfaces)
                 bool fRet = uiInterface.ThreadSafeQuestion(
-                    strLoadError + ".\n\n" + _("Do you want to rebuild the block database now?").translated,
+                    strprintf(_("%s.\n\nDo you want to rebuild the block database now?"), strLoadError),
+                    /*strLoadError + ".\n\n" + _("Do you want to rebuild the block database now?").translated,*/
```
:eyes:

hebasto replied:
It seems this line is outdated already ;)

hebasto replied:
Sorry for a mess. Fixed in the latest push 24cae76e6476f467095327ca9970ef1a784c629b.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1619,7 +1619,8 @@ bool AppInitMain(InitInterfaces& interfaces)
             if (!fReset) {
                 bool fRet = uiInterface.ThreadSafeQuestion(
-                    strLoadError + ".\n\n" + _("Do you want to rebuild the block database now?").translated,
+                    strprintf(_("%s.\n\nDo you want to rebuild the block database now?"), strLoadError),
```
Not sure if we want to translate `%s.\n\n`, maybe add an `operator+` or `operator<<` to bilingual_str, so that either `+` or `strprintf` can be used.

hebasto replied:
This line is outdated due to the following commits ;)

> ... maybe add an `operator+` or `operator<<` to bilingual_str, so that either `+` or `strprintf` can be used.

Agree. Good idea for the following PRs (as this PR does not touch `bilingual_str` itself).

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16224#issuecomment-623798715).

MarcoFalke commented on src/qt/bitcoingui.cpp:
```diff
@@ -1352,20 +1353,27 @@ void BitcoinGUI::showModalOverlay()
+
+    QString detailed_message;
+    if (message.original != message.translated) {
+        detailed_message = QObject::tr("Original message:") + "\n" + QString::fromStdString(message.original);
```
what is the point of translations again, if we show the original message anyway?

hebasto replied:
This translated title, `QObject::tr("Original message:")`, is intended for non-english interface.
The message itself is not translated, obviously.
That was an initial idea, but I'm open for changes ;)

MarcoFalke replied:
yes, what is the point of showing the translation *and* the original message. If the user understood the original message (in English), what is the point of showing the translation?

hebasto replied:
Translation - for a user.
Original message - for googling and referencing.

MarcoFalke replied:
I see. I guess fine with me, but at least the code could mention that.

hebasto replied:
Add some comments?

MarcoFalke replied:
jup

hebasto replied:
Done.

maflcko reviewed (commented):
Concept ACK. One question inline.

MarcoFalke commented on src/util/translation.h:
```diff
@@ -39,4 +39,9 @@ inline static bilingual_str _(const char* psz)
     return bilingual_str{psz, G_TRANSLATION_FUN ? (G_TRANSLATION_FUN)(psz) : psz};
 }
 
+inline static bilingual_str DoNotTranslate(std::string original)
```
in commit aa21c807af55f94867b11bab5e2e9a8160a3fc9d

style-question: Could this be called `Untranslated`, so that arbitrary strings can be passed in (even string that are impossible to translate, such as error strings returned from third party libraries).

hebasto replied:
Done.

hebasto commented:
@MarcoFalke 
Thank you for your review. Your comments have been addressed.

jonasschnelli commented:
Tested and looks good.
Would it make sense to also make the strings in qt/bitcoin.cpp translate the new way?
Things like `"Specified data directory \"%s\" does not exist"` or `"Error reading configuration file: %s\n"`

hebasto commented:
@jonasschnelli 

> Would it make sense to also make the strings in qt/bitcoin.cpp translate the new way?
> Things like `"Specified data directory \"%s\" does not exist"` or `"Error reading configuration file: %s\n"`

These string are not in `qt/bitcoin.cpp` but in `bitcoind.cpp` ;)
More details in #16366 by @MarcoFalke

hebasto commented:
Rebased after #17138 has been merged.

hebasto commented:
Rebased after #17070 has been merged.

hebasto commented:
Rebased after #17279 has been merged.

hebasto commented:
Rebased after #17260 has been merged.

hebasto commented:
Rebased after #16839 has been merged.

hebasto commented:
Rebased after #17304 has been merged.

hebasto commented:
Rebased after #16540 has been merged.

Sjors commented on src/httpserver.cpp:
```diff
@@ -175,7 +176,7 @@ static bool InitHTTPAllowList()
         if (!subnet.IsValid()) {
             uiInterface.ThreadSafeMessageBox(
-                strprintf("Invalid -rpcallowip subnet specification: %s. Valid are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24).", strAllow),
+                strprintf(_("Invalid -rpcallowip subnet specification: %s. Valid are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24)."), strAllow),
```
nit: make this `NoTranslate`. All relevant documentation is in English anyway.

laanwj replied:
Agree. This is a terrible message to translate.

hebasto replied:
Done.

MarcoFalke replied:
This has not yet been addressed

MarcoFalke replied:
I'd say we should leave all new translations for a later pull. You can blindly mark everything `Untranslated` for now

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16224#issuecomment-623798715).

Sjors reviewed (commented):
The end result ffed46515d556c0dbc8535548d0a485a62306106 looks good, lightly tested on macOS.

But the individual commits are too confusing atm, they would be easier to review if you:

1. the first commit 05b20f6 as is 
2. introduce `Untranslated` earlier, with just `{return original}`
3. the next commit moves `.translated` to `ThreadSafeMessageBox`
4. a commit where you introduce newly translated strings `_(`
5. a commit that drops the redundant `return false` from calls to `InitError` (although otherwise unrelated)
6. (maybe a few more things, to keep 7 focussed)
7. switch to bilingual_str

I don't mind if this PR adds or removes a few translations, if that makes things easier. I also suspect there's more places where we can add `Untranslated`, but this could be done in a followup.

laanwj commented:
ACK ffed46515d556c0dbc8535548d0a485a62306106

I like that this has a by-effect of introducing a type for GUI-facing messages versus internal ones. An explicit `Untranslated` instead of having `bilingual_str` implicitly constructed is a good idea. I also like that you put the translation-related stuff in a separate header.

hebasto commented:
Rebased after #17444 has been merged.

Also, [@Sjors's comment](https://github.com/bitcoin/bitcoin/pull/16224#discussion_r348510565) has been addressed.

hebasto commented:
@Sjors 

> But the individual commits are too confusing atm, they would be easier to review if you ...

Sorry, but I'd prefer to leave the commit order as is for the following reasons: (a) it helps me to keep this PR rebased; (b) other reviewers have ACKed already.
The intrinsic logic of each commit is: a functional change plus other changes to compile successfully.

> I also suspect there's more places where we can add `Untranslated`, but this could be done in a followup.

Totally agree with you.

hebasto commented:
Rebased after #16963 has been merged.

hebasto commented:
Rebased after #17943 has been merged.

Sjors commented:
Needs a rebase now that #17943 was reverted.

hebasto commented:
Rebased after #17754 and #17965 have been merged.

hebasto commented:
Rebased after #16702 and #17261 have been merged.

hebasto commented:
Rebased due to #17812.

hebasto commented:
Rebased due to #17812. Now fixed.

emilengler commented:
This is probably a low level error but running HEAD on german with `--datadir=/tmp/no` results in
![grafik](https://user-images.githubusercontent.com/12272949/76643704-e1044380-6555-11ea-8bf5-a41415bbb62f.png)

I miss the English message there

hebasto commented:
@emilengler 
> This is probably a low level error but running HEAD on german with `--datadir=/tmp/no` results in
> ![grafik](https://user-images.githubusercontent.com/12272949/76643704-e1044380-6555-11ea-8bf5-a41415bbb62f.png)
> 
> I miss the English message there

This is due to the following line:
https://github.com/bitcoin/bitcoin/blob/d600529cf44cde00212028f91c9ded5c4550d9ff/src/interfaces/node.cpp#L56

And I think this made intentionally after merging #16366. See some comments:
- https://github.com/bitcoin/bitcoin/pull/16224#issuecomment-514681825
- https://github.com/bitcoin/bitcoin/pull/16224#issuecomment-515566530

hebasto committed:
gui: Add detailed text to BitcoinGUI::message

hebasto committed:
Make ThreadSafe{MessageBox|Question} bilingual

hebasto committed:
Make InitError bilingual

hebasto committed:
doc: Do not translate technical or extremely rare errors

hebasto commented:
Rebased d600529cf44cde00212028f91c9ded5c4550d9ff -> 4bcdc3f4d9a5386b69142ca72ad500c64f9eab24 due to the conflict with #17737.

Sjors commented:
tACK 4bcdc3f

hebasto commented:
Rebased 4bcdc3f4d9a5386b69142ca72ad500c64f9eab24 -> 5ca0c77bae59f33dc122305d5f68077a70153af7 due to the conflict with #17954.

hebasto commented:
@ryanofsky 
@MarcoFalke 
@practicalswift 
@promag 
@laanwj 
@jonasschnelli 
@Sjors 
@emilengler 

Mind re-reviewing this PR to give it some--I hope final--(re)ACKs :)

MarcoFalke commented on src/wallet/load.cpp:
```diff
@@ -49,14 +49,15 @@ bool VerifyWallets(interfaces::Chain& chain, const std::vector<std::string>& wal
             return false;
         }
 
+        // TODO: error_string and warning_string should be bilingual strings.
```
I have a slight preference to get #18699 in first, which fixes this TODO

hebasto replied:
Agree.

maflcko reviewed (approved):
Concept ACK

Sjors commented:
Looks like it needs another rebase.

hebasto commented:
Rebased 5ca0c77bae59f33dc122305d5f68077a70153af7 -> cb041d501ac478a39ef73df86c3bce3b084cae18 due to the conflict with #15761.

Sjors commented:
re-tACK cb041d5

Can you update the PR description with an example that actually works? E.g. `bitcoin-qt -lang=nl -uacomment="\\"`

hebasto commented:
> re-tACK [cb041d5](https://github.com/bitcoin/bitcoin/commit/cb041d501ac478a39ef73df86c3bce3b084cae18)
> 
> Can you update the PR description with an example that actually works? E.g. `bitcoin-qt -lang=nl -uacomment="\\"`

Thank you! Updated.

hebasto commented:
Rebased cb041d501ac478a39ef73df86c3bce3b084cae18 -> https://github.com/bitcoin/bitcoin/commit/0ea49f4c28ca9dd7cf5ba81c66041d2c2f7b2e92 due to the conflict with #16528.

Sjors commented:
utACK 0ea49f4c28ca9dd7cf5ba81c66041d2c2f7b2e92 rebase

hebasto commented:
Rebased 0ea49f4c28ca9dd7cf5ba81c66041d2c2f7b2e92 -> 1078fad9c517059d3752d01cdca189994fac0d4b due to the conflicts with #16426 and #18727.

Sjors commented:
Travis didn't run. Make sure your account didn't accidentally get flagged.

hebasto commented:
> Travis didn't run. Make sure your account didn't accidentally get flagged.

Verified. It seems ok.

Sjors commented:
Ok, maybe just try another rebase?

hebasto commented:
> Ok, maybe just try another rebase?

Rebased 1078fad9c517059d3752d01cdca189994fac0d4b -> 101ff19cdc0743c48ec76d9a39855be632de7956.

MarcoFalke commented:
Needs rebase, but diff should be slightly smaller because of the merge of #18699

hebasto commented:
Rebased 101ff19cdc0743c48ec76d9a39855be632de7956 -> 5ff042d4955ab0f8a0dd8ed2522cf0da17f5b00d due to the conflict with #18699.

Sjors commented:
re-tACK 5ff042d

hebasto committed:
util: Cleanup translation.h

hebasto commented:
Updated 5ff042d4955ab0f8a0dd8ed2522cf0da17f5b00d -> 18bd83b1fee2eb47ed4ad05c91f2d6cc311fc9ad:

- this PR does not add any new error message translations
- added "util: Cleanup translation.h" commit to address @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18699#discussion_r419597778) from #18669

Sjors commented:
re-tACK 18bd83b1fee2eb47ed4ad05c91f2d6cc311fc9ad

MarcoFalke commented:
ACK 18bd83b1fee2eb47ed4ad05c91f2d6cc311fc9ad 🐦

Thanks for finally making the debug log of international users readable. I
found it painful to translate the debug log in bug reports back to English.

I checked that no new translations are added, but did not run the GUI.

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 18bd83b1fee2eb47ed4ad05c91f2d6cc311fc9ad 🐦

Thanks for finally making the debug log of international users readable. I
found it painful to translate the debug log in bug reports back to English.

I checked that no new translations are added, but did not run the GUI.
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiI8wwAnm6ab4DxH44gY+MYdYKbbQe+Qayc2wuMw1wvpNfYdJDZ2xK2N5ZWCIoc
grA4eVeluZggMf30d8pKP+HFofQVugYFyTB/lG/9PnSl//WAgezClqkxnW1YE0pm
xfVfKC1f7RBGswfcm+Rrq7m0LfzWEjoGgCCbcjQm86kAP+Mt8phmjQRNc43FycFJ
dLO7DFZMiqjSIhLi4bbSqiynUXuTY2tMbjv4e23gHWKFcnvndBE3QjfirsRlcp4f
Ly5lvEFCABOdnMQ0K8/9bSn2XqHwWyZxXJ+Yj1twkCchcjmcnHTfAmTJXs0Vahqd
vapzAw2X0dk6x8JWZcu+am7rFSzZu+tAHWnDC1DgjgsnDZMMTI4rjPbtBP9X01ka
C6SjRDt91NZdm2x+jXrHKEhmQdQ3FDVU2hGb6g3cjxe/RJAyxZ3W5ZTTYqvQXOjh
992J7pQOtQjQHovjFWyQ4WSxlTXPd7PFhZ1NyOJE2UCJD8rdUtav2nxzbh0q3SX2
Seu4nsii
=xXpH
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `1fd623baf008aff28a965813df642d958b9fc77a22a30025ff01d55db57e248e  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401081fd623baf008aff28a965813df642d958b9fc77a22a30025ff01d55db57e248ef010fa2800e2d83d44c34b8a5d06fb660cc008fff01001d7f5a5ac95a6bee6bc8bfe297ab91a08f1045eb58095f0083807fec70ac4a3900083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0100678d1371c3718f3c2d35d6f88d2343708f1045eb58094f008f4753e2ee04475320083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c980a30d30064f94ffd4de6f3cf61dde08f1045eb58095f0089981896457344bed0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010fa4ae702853347a32f5d08171e93114c08f1045eb58095f0082962d916cfe48e0f0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
